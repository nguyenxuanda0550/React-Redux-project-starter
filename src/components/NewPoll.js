import { useState } from 'react';
import { connect } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import { handleCreateQuestion } from '../actions/polls';

const NewPoll = ({ dispatch }) => {
  const navigate = useNavigate();
  const [options, setOptions] = useState({
    firstOption: '',
    secondOption: '',
  });

  const onSubmit = (e) => {
    e.preventDefault();
    if (options.firstOption === '' || options.secondOption === '') {
      alert('Please enter both options');
      return;
    }
    dispatch(handleCreateQuestion(options.firstOption, options.secondOption));
    navigate('/');
  };

  return (
    <div className="container mt-4">
      <h1 className="text-center mb-4">New Poll</h1>
      <form className="mx-auto" style={{ maxWidth: '600px' }} onSubmit={onSubmit}>
        <div className="mb-3">
          <label
            htmlFor="firstOption"
            className="form-label"
            data-testid="firstOptionLabel"
          >
            First Option
          </label>
          <input
            value={options.firstOption}
            onChange={(e) =>
              setOptions({ ...options, firstOption: e.target.value })
            }
            type="text"
            name="firstOption"
            data-testid="firstOption"
            className="form-control"
          />
        </div>
        <div className="mb-3">
          <label
            htmlFor="secondOption"
            className="form-label"
            data-testid="secondOptionLabel"
          >
            Second Option
          </label>
          <input
            value={options.secondOption}
            onChange={(e) =>
              setOptions({
                ...options,
                secondOption: e.target.value,
              })
            }
            type="text"
            name="secondOption"
            data-testid="secondOption"
            className="form-control"
          />
        </div>
        <button
          type="submit"
          className="btn btn-primary w-100"
          data-testid="submit-poll"
        >
          Add Poll
        </button>
      </form>
    </div>
  );
};

export default connect()(NewPoll);
