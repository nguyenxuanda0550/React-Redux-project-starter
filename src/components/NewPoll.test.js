import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import NewPoll from './NewPoll';
import { configureStore } from '@reduxjs/toolkit';
import reducer from '../reducers';

// Configure Redux store with the provided reducer
const store = configureStore({ reducer });

describe('NewPoll Component', () => {
  // Test to check if the NewPoll component renders correctly
  test('should render the component', () => {
    // Render NewPoll component wrapped with Redux Provider and React Router
    const view = render(
      <Provider store={store}>
        <BrowserRouter>
          <NewPoll />
        </BrowserRouter>
      </Provider>
    );

    // Assert that the component is defined and matches the snapshot
    expect(view).toBeDefined();
    expect(view).toMatchSnapshot();
  });

  // Test to check if all elements in the NewPoll component are displayed correctly
  test('should display all elements', () => {
    // Render NewPoll component wrapped with Redux Provider and React Router
    render(
      <Provider store={store}>
        <BrowserRouter>
          <NewPoll />
        </BrowserRouter>
      </Provider>
    );

    // Select elements by their test IDs
    const firstOptionLabelElement = screen.getByTestId("firstOptionLabel");
    const firstOptionInputElement = screen.getByTestId('firstOption');
    const secondOptionLabelElement = screen.getByTestId('secondOptionLabel');
    const secondOptionInputElement = screen.getByTestId('secondOption');
    const submitButtonElement = screen.getByTestId('submit-poll');

    // Assert that the text content of labels and button is correct
    expect(firstOptionLabelElement.textContent).toBe('First Option');
    expect(secondOptionLabelElement.textContent).toBe('Second Option');
    expect(submitButtonElement.textContent).toBe('Add Poll');

    // Simulate user input and assert that input values are updated correctly
    fireEvent.change(firstOptionInputElement, { target: { value: 'Texas' } });
    fireEvent.change(secondOptionInputElement, { target: { value: 'New Hampshire' } });
    expect(firstOptionInputElement.value).toBe('Texas');
    expect(secondOptionInputElement.value).toBe('New Hampshire');
  });
});
