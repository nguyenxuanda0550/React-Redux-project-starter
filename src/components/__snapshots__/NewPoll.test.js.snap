// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewPoll Component should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container mt-4"
      >
        <h1
          class="text-center mb-4"
        >
          New Poll
        </h1>
        <form
          class="mx-auto"
          style="max-width: 600px;"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              data-testid="firstOptionLabel"
              for="firstOption"
            >
              First Option
            </label>
            <input
              class="form-control"
              data-testid="firstOption"
              name="firstOption"
              type="text"
              value=""
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              data-testid="secondOptionLabel"
              for="secondOption"
            >
              Second Option
            </label>
            <input
              class="form-control"
              data-testid="secondOption"
              name="secondOption"
              type="text"
              value=""
            />
          </div>
          <button
            class="btn btn-primary w-100"
            data-testid="submit-poll"
            type="submit"
          >
            Add Poll
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container mt-4"
    >
      <h1
        class="text-center mb-4"
      >
        New Poll
      </h1>
      <form
        class="mx-auto"
        style="max-width: 600px;"
      >
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            data-testid="firstOptionLabel"
            for="firstOption"
          >
            First Option
          </label>
          <input
            class="form-control"
            data-testid="firstOption"
            name="firstOption"
            type="text"
            value=""
          />
        </div>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            data-testid="secondOptionLabel"
            for="secondOption"
          >
            Second Option
          </label>
          <input
            class="form-control"
            data-testid="secondOption"
            name="secondOption"
            type="text"
            value=""
          />
        </div>
        <button
          class="btn btn-primary w-100"
          data-testid="submit-poll"
          type="submit"
        >
          Add Poll
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
